Add rubocop code analyser

clarissalimab opened the pull request:
Closes #488.

To see the code analysis locally, run:

```bash
rubocop
```

wwahammy commented:
The checks seem to be failing for both pronto and the main_build.

wwahammy commented on .github/workflows/pronto.yml:
```diff
@@ -0,0 +1,26 @@
+name: Pronto
+on: ['push', 'pull_request']
+
```
We should limit this job to only run when ruby related files have changed. Perhaps use the same conditions as main_build uses?

Alternatively, this could be put into the main build.

clarissalimab replied:
Great! I'll add to the main_build

clarissalimab replied:
Actually, I think it would be better to run in another job, because we'd probably want to know when both are failing.

wwahammy replied:
In theory, I agree. Unfortunately, it takes a really long time for new jobs to boot as it relates to loading the cache and installing any missing packages. Ideally, everything that can run in parallel should but it's not really practical.

wwahammy commented on .rubocop.yml:
```diff
@@ -0,0 +1,20 @@
+AllCops:
+  NewCops: enable
+  Exclude:
+    - db/**/*
+    - tmp/**/*
+    - Gemfile
+    - Gemfile.lock
```
Doesn't this imply that all the old files would be  tested if we ran rubocop?

clarissalimab replied:
Sorry, why? This `Exclude` is ignoring those files.

wwahammy commented on Gemfile:
```diff
@@ -3,18 +3,18 @@
 source 'https://rubygems.org'
 
 ruby '2.6.6'
-gem 'rails', '~> 6.1.3'
-gem 'jbuilder', '~> 2.10'
```
Make sure not reorder this file, especially things related to rails. Order matters in `Gemfile`s. There's a definitely formatting standards to implement in this file but reordering the imports is not one of them.

wwahammy reviewed (changes requested):
Good stuff here but there are some changes you'll need to make.

I'm also not entirely sold on using pronto. Have you used it before?

clarissalimab commented:
I have used pronto before. It's not essential but helps with not having to run

`git diff --name-only --no-commit-id $(git merge-base origin/main rubocop) | xargs rubocop --force-exclusion`

because we don't want to run on files that were not changed. And we can use it for reek too if we add it later.

wwahammy commented:
How much time do we save on each build if we run pronto instead of rubocop?

clarissalimab commented:
I don't think it saves time, it helps to run the code analysis tools only on the relevant files - which in our case means only running on the modified files instead of in the entire project. Rubocop does not have that configuration, but we could always run only the command I mentioned before to filter changed files between `main` and the current branch.

wwahammy commented:
I'm still not sure we're on the same page regarding limiting that we run rubocop only on what changed. What's the advantage of that?

wwahammy commented:
I get it if it saves a bunch of time but if the time difference isn't significant what are we gaining?

clarissalimab commented:
If we use pronto, we exclude files that currently are in the codebase in the analysis. Rubocop itself does not have a configuration to ignore files that already exist.

If we simply run rubocop, we get:
![image](https://user-images.githubusercontent.com/15739610/108783678-a6b86a00-754c-11eb-9867-48b1bf945337.png)

* The first workaround for this would be using pronto:
![image](https://user-images.githubusercontent.com/15739610/108783834-eaab6f00-754c-11eb-8a2c-412c503d0f3e.png)
In this case, it checked only the changed lines.

* Or running `git diff --name-only --no-commit-id $(git merge-base origin/main rubocop) | xargs rubocop --force-exclusion`:
![image](https://user-images.githubusercontent.com/15739610/108784218-aa98bc00-754d-11eb-9760-42b3988d7c78.png)

In this case, it checks the entire file instead of just the lines I've changed, while pronto only checks the lines that I've changed.

That's not our case now, but Pronto would also help if we wanted to add other tools, like `reek` that I mentioned on the issue.

wwahammy commented:
Couldn't we use exclude in the robocop config to exclude the current files?

wwahammy commented:
We actually do something similar in ESLint to exclude old files.

On Mon, Feb 22, 2021, 5:48 PM Eric Schultz <eric@commitchange.com> wrote:

> Couldn't we use exclude in the robocop config to exclude the current files?
>

wwahammy commented:
https://www.github.com/houdiniproject/houdini/tree/main/.eslintignore

clarissalimab commented:
We could, but I thought we'd want to see the warnings if we modified the current files, no? We want rubocop to run only on new files?

wwahammy commented:
My thinking was we'd only run it on new files and we would slowly add old files over time.

My worries about only running on changes is that it tends to be a pain to set up and I think two people with identical file trees could run it on their systems and their rubocop results could vary. But maybe I'm overthinkinf that last part.

clarissalimab commented:
No problem, I can remove **pronto** and use only **rubocop** then, but doesn't adding the files slowly take the same effect in the end, except that it's done manually?

wwahammy commented:
> No problem, I can remove **pronto** and use only **rubocop** then, but doesn't adding the files slowly take the same effect in the end, except that it's done manually?


Pretty much, we get the same result at the end which is getting code formatted to our standard going forward.

clarissalimab commented:
It's ready for review now :smiley:

wwahammy commented on .github/workflows/ruby.yml:
```diff
@@ -68,4 +68,6 @@ jobs:
-        run: bin/rails spec
\ No newline at end of file
+        run: bin/rails spec
+      - name: rubocop
```
Please reorder these. It's better to have rubocop fail first because it's quicker.

wwahammy reviewed (changes requested):
Once minor change and then I think it's ready to go!

wwahammy reviewed (approved):
:+1: :fireworks:
